src/udp_packet.cpp

```diff
@@ -0,0 +1,117 @@
+#include <stdexcept>
+#include <sstream>
+
+#include "udp_packet.h"
+
+namespace
+{
+	const int UDP_HEADER_SIZE = 8;
+}
+
+UdpPacket::UdpPacket(uint16_t sourcePort, uint16_t destinationPort,
+					 const std::vector<uint8_t> &payload) :
+	m_sourcePort(sourcePort),
+	m_destinationPort(destinationPort),
+	m_checksum(0),
+	m_payload(payload)
+{
+	m_packetLength = UDP_HEADER_SIZE + m_payload.size();
+}
+
+UdpPacket::UdpPacket(const std::vector<uint8_t> &buffer)
+{
+	auto bufferSize = buffer.size();
+	if (bufferSize < UDP_HEADER_SIZE)
+	{
+		std::stringstream stream;
+		stream << "UDP header is " << UDP_HEADER_SIZE << " bytes, but buffer "
+			   << "is only " << bufferSize << " bytes";
```

Comment:
Yeah my indentation is all over the place due to [this age-old bug](https://bugreports.qt.io/browse/QTCREATORBUG-7994). Should be better now.

As for the stream operator, I just feel like it lends itself to better readability (when the indentation is correct).